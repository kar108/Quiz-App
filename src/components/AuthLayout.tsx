import { useContext } from "react";
import { Outlet, Navigate } from "react-router-dom";
import { AppContext } from "../App";


const AuthLayout = () => {
  const[appData]=useContext(AppContext);
  console.log(appData==={ answers: {} });
  const  areQuestionsNotAnswered=false;
  return areQuestionsNotAnswered ? (
    <Navigate to="/" replace />
  ) : (
    <Outlet />
  );
};

export default AuthLayout;