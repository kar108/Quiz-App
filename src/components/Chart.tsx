import { color } from '@mui/system';
import React from 'react';
import { PieChart, Pie, Tooltip } from 'recharts';
import "./Chart.css";

const Chart = (data:any) => {
    const chartData=data.data;
    return (
    <div className='chart'>
        <PieChart width={400} height={400}>
        <Pie
          dataKey="value"
          isAnimationActive={true}
          data={chartData}
          cx="50%"
          cy="50%"
          outerRadius={100}
          fill="#24344d"
          label
        />
        <Tooltip />
      </PieChart>
    </div>
    );
};

export default Chart;