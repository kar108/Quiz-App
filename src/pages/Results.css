table {
    border-collapse: collapse;
    margin-top: 1rem;
    width: 100%;
  }
  
  tr {
    display: flex;
  }
  
  table a {
    text-decoration: none;
    color: white;
  }
  
  th {
    background-color: #24344d;
    color: var(--backgroundColor);
  }
  
  th,
  td {
    padding: 10px;
    text-align: center;
    flex: 1;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }

  .hide{
       visibility:collapse;
  }
  

