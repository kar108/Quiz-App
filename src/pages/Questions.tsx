import { useEffect, useState } from "react";
import QuestionCard, { QuestionType } from "../components/QuestionCard";
import questions from "../questions.json";
import { Box, Button, Typography, CircularProgress } from "@mui/material";
import { useNavigate } from "react-router-dom";
import { useContext } from "react";
import { AppContext } from "../App";
import { Paper} from "@mui/material";
import { hover } from "@testing-library/user-event/dist/hover";
import AuthLayout from "../components/AuthLayout";

export const TOTAL_QUESTIONS = questions.length;

export const Questions = () => {
  const [currentQuestion, setCurrentQuestion] = useState(questions[0].id);
  const [isLoading, setIsLoading] = useState(true);
  const navigate = useNavigate();
  const [appData] = useContext(AppContext);

  useEffect(() => {
    setTimeout(() => setIsLoading(false), 1000);
  }, []);

   function handleNextButtonClick(index: number) {
    if (index === TOTAL_QUESTIONS - 1) {
      navigate("/results");
    } else {
      setCurrentQuestion((prev) => prev + 1);
    }
  }

  return (
    <Box
      height="100vh"
      display="flex"
      justifyContent="center"
      alignItems="center"
    >
      <Box>
      </Box>
      {isLoading ? (
        <CircularProgress />
      ) : (
         questions.map(
          (question, index) =>
            currentQuestion === question.id && ( 
              <Box key={question.id}>
                <Paper sx={{height:"50px",width:"500px",padding:"10px 20px" ,display:'flex',alignContent:"center",justifyContent:"space-between",marginBottom:5}}>
                  Question Status:
                  {questions.map((question,index)=> (<Button sx={{backgroundColor:currentQuestion === index+1 ? "yellow":appData.answers[index+1]?.value && appData.answers[index+1].value.length!==0 ? "green":"#34495e",color:"black","&:hover":{backgroundColor:"yellow"}}} onClick={()=>{setCurrentQuestion(index+1);}}>{index+1}</Button>))}
                </Paper>
                <Typography variant="h3">Question: {index + 1}</Typography>
                <Box
                  height="500px"
                  display="flex"
                  flexDirection="column"
                  justifyContent="space-between"
                >
                  <QuestionCard
                    id={question.id}
                    question={question.question}
                    questionType={question.questionType as QuestionType}
                    answers={question.answerOptions}
                  />
                  <Box
                    sx={{
                      display: "flex",
                      alignItems: "center",
                      justifyContent: "space-between",
                    }}
                  >
                    <Button
                      disabled={index === 0}
                      onClick={() => setCurrentQuestion((prev) => prev - 1)}
                    >
                      Previous
                    </Button>

                    <Button onClick={() => handleNextButtonClick(index)}>
                      {index === TOTAL_QUESTIONS - 1 ? "Submit" : "Next"}
                    </Button>
                  </Box>
                </Box>
              </Box>
            )
        )
      )}
    </Box>
  );
};

export default Questions;