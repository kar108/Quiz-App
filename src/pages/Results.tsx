import { join } from "path";
import React, { useContext } from "react";
import { AppContext } from "../App";
import questions from "../questions.json";
import Chart from "../components/Chart";
import "./Results.css";
import { Typography } from "@mui/material";
import Questions from "../questions.json";


const n=Questions.length;


const Results = () => {
  let CorrectAns=0;
  let WrongAns=0;
  const [appData] = useContext(AppContext);
  const res=questions.map((question,index)=>{return question.answerOptions.map((choice,index)=>{return choice.isCorrect?choice.option:null})});
  const ans = res.map((items)=>{return items.filter((item)=>item!==null)});
  return (
    <div>
      <Typography variant="h3" sx={{ fontSize:25, display:"flex",alignItems:"center",justifyContent:"center",marginTop:"1rem"}}>
      QUIZ - RESULT
      </Typography>
       <div>
         <table>
           <thead>
             <tr>
               <th>Index</th>
               <th>Answer</th>
               <th>Expected-Answer</th>
               <th>Remark</th>
             </tr>
           </thead>
           <tbody>
           {Object.entries(appData.answers).map(([currentIndex, answer]: any) => (
             <tr key={currentIndex}>
               <td>{currentIndex}</td>
               <td>{answer.value}</td>
               <td>{ans[currentIndex-1]}</td>  
               <td>{answer.value!=null?(answer.value==ans[currentIndex-1]?"Correct":"Incorrect"):"Incorrect"}</td>
               <span className="hide">{answer.value==ans[currentIndex-1]?CorrectAns+=1:null}</span>
             </tr>
           ))}
           </tbody>
         </table>
      </div>
      <Chart data={[{name:"CorrectAns",value:CorrectAns},{name:"WrongAns",value:n-CorrectAns}]}/>
    </div>
  );
};


export default Results;