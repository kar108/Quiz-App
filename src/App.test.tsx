import React, { ChangeEvent } from 'react';
import { render, screen } from '@testing-library/react';
import App, { AppContext } from './App';
import {initContextData} from "./App";
import Questions, { TOTAL_QUESTIONS } from './pages/Questions';
import InputDetailsForm from './pages/InputDetailsForm';
import { Radio } from '@mui/material';
import Results from './pages/Results';
import { Router, Routes } from 'react-router-dom';
import Chart from './components/Chart';
import RadioButtons from './components/buttons/RadioButtons';
import AuthLayout from './components/AuthLayout';
import CheckboxOptions from './components/CheckboxOptions';
import RadioOptions from './components/RadioOptions';



const initdata={ answers: {} };

describe("unit testing basics", () => {  
  test("verifying initdata", () => {
    expect(initContextData).toStrictEqual(initdata);
  });
});

describe("unit testing basics", () => {  
  test("verifying number of questions", () => {
    expect(TOTAL_QUESTIONS).toStrictEqual(4);
  });
});

describe("test Item component", () => {
  test("render item component", () => {
    const appData={answers:{1:{type:"Radio",value:["Every 15 days"]},},};
    render(<AppContext.Provider value={[appData]}>
      <Results/>
    </AppContext.Provider>);
    const title=screen.getByText("QUIZ - RESULT")
    expect(title).toBeInTheDocument();
  });
});

describe("test Item component", () => {
  test("render item component", () => {
      <Chart data={[{name:"CorrectAns",value:1},{name:"WrongAns",value:2}]}/>
  });
});

describe("test Item component", () => {
  test("render item component", () => {
      <RadioButtons label={''} buttons={[]}/>
  });
});

describe("test Item component", () => {
  test("render item component", () => {
    const appData={answers:{1:{type:"Radio",value:["Every 15 days"]},},};
    render(<AppContext.Provider value={[appData]}>
      <AuthLayout/>
    </AppContext.Provider>);
  });
});

describe("test Item component", () => {
  test("render item component", () => {
    const fnt=jest.fn();
    render(
      <CheckboxOptions options={[]} handleChange={fnt}/>);
  });
});

describe("test Item component", () => {
  test("render item component", () => {
    const fnt=jest.fn();
    render(
      <RadioOptions options={["Every 15 days"]} value={'true'} handleChange={function (e: ChangeEvent<HTMLInputElement>): void {
        throw new Error('Function not implemented.');
      } }/>
      );
      screen.debug()
      const title=screen.getByText("Every 15 days")
      expect(title).toBeInTheDocument();
  });
});



test("snapshot testing", () => {
  const fnt=jest.fn();
  const { container } = render(<CheckboxOptions options={[]} handleChange={fnt} />);
  expect(container.firstChild).toMatchSnapshot();
});