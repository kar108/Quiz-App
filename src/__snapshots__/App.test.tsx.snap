// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
<div
  class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
/>
`;
